### ISSUE TITLE
Release swarm for continuous delivery

### ISSUE THREAD

**Issue by oskarth** [status-im/ideas#64]
## Preamble

    Idea: PROC064
    Title: Release swarm for continuous delivery
    Status: Draft
    Created: 2017-12-27
    Started: 2018-01-04
    Requires (*optional): -
    Replaces (*optional): -

## Summary

A swarm dedicated to making sure we do regular releases, so we keep improving the app for end users and other core contributors aren't needlessly distracted by release work. A cadence for the team.

A secondary goal is to improve the release process when it comes to automation, changelogs, testing, and having more people who can do this.

## Swarm Participants

- Lead Contributor: @oskarth (10h/w)
- Testing & Evaluation: @anna (20h/w) @asemiankevich (20h/w)
- Clojure Contributor: @roman (10h/w)
- QA Automation Contributor: @antdanchenko (15h/w January)
- Go Contributor: adambabik (10h/w+)
- Community/marketing: @andytudhope (5h/w)
- Community/marketing: @pablanopete (5h/w)

- PM: @chad? (TBC)
- Clojure Contributor: One more?
- Testing & Evaluation: ?

Swarm size: estimated 4-10 people.

## Product Overview

Regular releases are like exercise, it's easy to get out of shape. In the last six months or so we have been quite bad at this. Each release has dragged out on time, and we have compromised release dates for scope creep. Getting back onto a regular release schedule is in the interest of many groups of people, as can be seen in the user stories below.

### User stories

1. As a developer, I want to not have to worry about when a release is going to happen and just keep working so I can be more productive.

2. As a swarm leader, I want releases to be in some cadence so I can somewhat plan iterations in relating to releases.

3. As a PM, I want releases to happen on a clockwork so I can see clear progress being made on a reasonable time frame.

4. As someone responsible for releases, I want to have an established procedure for releases, with as much automation and instructions as possible, so I can either do less work or someone else can do this, and I can focus on more important things.

5. As a tester, I want to have a process for when to do thorough testing and establish procedures for quickly testing release candidates.

6. As an end user, I want to use the latest version of Status so I can get benefits and give feedback early.

7. As a developer, I want to have regular releases so I feel motivated by the prospects of things I have worked on being included.

This is a process and a cross-functional team we are introducing. Once a few releases have gone by we expect this process to be considered "installed" and the group will continue to exist.

### Requirements & Dependancies

- 1) Swarm autonomy to decide on scope of release, including when to cut release branch, revert commits, etc. Releases aren't based on scope but on dates, i.e. scopes will likely be cut.

- 2) Account details for Google Play Store and Apple equivalent.

- 3) Bug hunting day at appropriate timing to release branch.

- 4) Release branch cut from develop ~1w before release and only bug fixes are cherry-picked to this.

- 5) Known availablity of dev/QA since iterations are short and we currently have some SPOF.

### Minimum Viable Product

Goal Date: 2017-01-14 (0.9.13)
Actual end date: 2017-01-30 (2w delayed due to regressions)
Description: Main deliverable: First release. Additionally: Cut release branch one week before on appropriate day. Oskar doing release with help from Roman and Anna as they have both done this. Develop rudimentary functional spec, i.e. supported user stories that are tested. Parallel track is decentralizing testing of these releases too. TBD and up to Anna. Start writing checklist for wiki;
start thinking about automation; setup Jenkins RC build; release notes to hutch/Chad.

END: All done except: Develop rudimentary functional spec, i.e. supported user stories that are tested. Parallel track is decentralizing testing of these releases too.

### Iteration 1

Goal Date: 2017-02-15

Description: Release 0.9.14 out and cut by Eric (3rd person). Fastlane Jenkins setup pipeline (auto build Android/iOS).

TBD specific commitment around: CD Anton auto testing part of pipeline.

TBD specific commitment around: QA Nightly / auto testing.

### Iteration 2
Goal Date: 2017-02-14 (0.9.15), or +2w iteration 1

Description: TBD. Once these 3 releases have been done, hopefully reasonably according to plan, this idea can be put as closed and swarm continues to exist in background. At some point we might want to get a release engineer to off-load some routine work.

Testing Days required: N/A (part of process)

## Success Metrics

- 1) Releases happen on a known cadence, like a clockwork.

- 2) A process involving as few people as possible to ensure most core contributors keep shipping.

- 3) Automation, checklists etc which minimizes work required and uncertainty for a release.

- 4) Release work is largely a decentralized capability and doesn't depend on a SPOF (Roman, Anna).

## Supporting Role Communication

- Marketing / PM / Community.

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

**Comment by oskarth** [comment]
I pledge 10h a week for 4-6 weeks starting from January. Swarm channel: #64-release.

**Comment by oskarth** [comment]
More generally we want to introduce a form of Release Engineering or Continuous Delivery at Status.

**Comment by adambabik** [comment]
Happy to pledge 10h (or more if needed) from the Go Team if needed.

From my experience, what works the best is an automated CI/CD process that is triggered by git events like a merge or new tag. Custom slack or Github bots are problematic and needs a lot of maintenance effort.

**Comment by andytudhope** [comment]
I'll pledge 5h/week from community perspective seeing as I am the one managing the App Store etc.

**Comment by pablanopete** [comment]
Glad to help with social posting ~5hr whenever needed, please include me in any discussions had on this swarm going forward in pings as I'd like to know what our new cadence will be.

**Comment by antdanchenko** [comment]
I pledge 15h a week for 4+ weeks starting from January 1st.

**Comment by annadanchenko** [comment]
20 hrs a week starting from January 4th

**Comment by oskarth** [comment]
MVP defined and key people committed - let the swarm commence!

**Comment by adambabik** [comment]
I'd like to describe how it looks like from status-go developer perspective and how it can be improved.

Currently, when a new PR is merged into status-go, we have no idea and no means to verify it easily if the Status app is still working correctly. In such a case, I think GoTeam should be actually responsible for that, not ClojureTeam. I have been thinking about an automated way of opening a new PR in status-react with an updated status-go version so that a Go developer can download status-react PR and build it at least locally. (Of course, we also need a mechanism that will not open such PRs in case of updating a README or something like that. Also, we can fall back to do it manually for the time being but an instruction is required.)

In status-go, we also need a master branch. I think it's ok if status-go develop branch is broken for a bit as a result of merging a PR. However, the master branch should be always working with a particular status-react version. Master should be actually used to do a release.

I would define two action items to get it going for now:
1. Prepare a manual instruction (in status-go wiki) how to build the Status app locally with an updated status-go version on simulator and real-device for iOS and Android,
2. Start updating the master branch of status-go for versions tested by QA. I imagine we can prepare PRs in status-react with an updated status-go version and if we get 👍  from QA, a master branch in status-go can be updated.

**Comment by chadyj** [comment]
I'm really looking forward to this swarm evolving to the point that our releases happen seamlessly and like clockwork!

One tweak I'd suggest as we head towards beta is making sure we satisfy Apple's [guideline](https://developer.apple.com/app-store/review/), performance, and technical requirements as they are known to be incredibly strict. In the past I've seen these issues push back releases for weeks or even months, and in our case it is better to be aware of these issues now rather than block a release down the line.

Fortunately this is easy to do through the [App Store submission process](https://help.apple.com/itunes-connect/developer/#/dev34e9bbb5a). I propose that when we have a build that is sufficiently complete we submit (with the release criteria set to `manually release this version`/`pending developer release` so it isn't actually published). If the app is rejected we will be able to prioritize these issues and have it fixed for the next release.

**Comment by oskarth** [comment]
Good idea @chadyj.

Release 0.9.13 delayed a few days due to major regressions since last release. Total of 36 commits specific to this release, ideally this should be low single-digits (diff config, no regressions). ETA with no additional info is +1w, i.e. Thursday.

**Comment by adambabik** [comment]
@oskarth when 0.9.13 is done, please inform here which status-go version (git commit SHA) was used. I will update status-go master branch and create a tag so that we can track release info in status-go as well.

**Comment by oskarth** [comment]
@feuGeneA @adambabik That tag isn't final. It is moving as we are finalizing release, which isn't done yet. Will update as soon as release is final. It is very likely it will stay as `develop-gd71c66a2` though (https://github.com/status-im/status-react/blob/release/0.9.13/modules/react-native-status/ios/RCTStatus/pom.xml#L28), so you can assume this unless something new comes up.

**Comment by adambabik** [comment]
@oskarth this is not a good practice to change a tag after it's created. Maybe you could use a release branch like `release/0.9.13`. That would be more transparent what's going on.

**Comment by oskarth** [comment]
Why are we talking about implementation details in a meta ideas issue? The tag stuff was needed to solve a specific problem related to how artifacts are tagged for QA. It's not ideal to move tags but it's also not a problem. The tagged version is already a RC and we already use release branches as well. The process for doing this has already been solved and automated, and unless you are particularly blocked by it (have no idea why that'd be the case) it's a non-issue.

In general, we have a project board fir these types of detailed stuff here:
https://github.com/status-im/status-react/projects/10

As well as a Slack channel (#64-release). Let's leave this issue open for more general details and not get bogged down in technical details.

**Comment by oskarth** [comment]
Hey everyone! 0.9.13 is now released: https://blog.status.im/status-alpha-0-9-13-release-notes-e1f918b75368

Thanks everyone who made the release happen. Extra kudos to @anna for all her tireless testing these last few weeks, and also especially to @janherich, @dmitryn and @yenda who went the extra mile in terms of fixing _several_ regressions that were blocking release! Even though they had swarm work, like everyone else, they still took it upon themselves to do everything necessary for us to to get this release out there. This is absolutely critical for Status Core as a product to succeed in delivering a high-quality product, so I just want to re-iterate my thanks for this generally thankless task

---

@adambabik @feuGeneA confirming that status-go version is the same as posted earlier.

---

Release 0.9.14 branch started. ETA ~2w. @yenda is release cutter this time. Will update with iteration goals soon.

**Comment by oskarth** [comment]
Meeting x2 weekly, today's notes: https://docs.google.com/document/d/1IUaMbz5rR0ce0laRRA-c97gFXEb6bfWrL_fyJTF3B4w/edit#heading=h.c7fy5an8xcrq

---

MVP iteration was:

Goal Date: 2017-01-14

Actual goal date: 2017-01-30 (2w delayed due to regressions)

Description: Main deliverable: First release. Additionally: Cut release branch one week before on appropriate day. Oskar doing release with help from Roman and Anna as they have both done this. Develop rudimentary functional spec, i.e. supported user stories that are tested. Parallel track is decentralizing testing of these releases too. TBD and up to Anna. Start writing checklist for wiki; start thinking about automation; setup Jenkins RC build; release notes to hutch/Chad.



All done except: Develop rudimentary functional spec, i.e. supported user stories that are tested. Parallel track is decentralizing testing of these releases too.



---



Iteration 1

Goal Date: 2017-02-15

Description: Release 0.9.14 out and cut by Eric (3rd person). Fastlane Jenkins setup pipeline (auto build Android/iOS).



TBD specific commitment around: CD Anton auto testing part of pipeline.

TBD specific commitment around: QA Nightly / auto testing.

**Comment by oskarth** [comment]
Updated Goal Date for 0.9.14 given that we have 9 open issues and it is 3 days to 15th: https://github.com/status-im/status-react/search?utf8=%E2%9C%93&q=is%3Aopen+label%3Arelease&type=Issues

**Comment by oskarth** [comment]
Iteration 1 DONE
Goal Date: 2017-02-15
Actual goal date: 2017-02-22
Description: Release 0.9.14 out and cut by Eric (3rd person). Automated test: Successful build of PR will trigger automated tests on each PR and post results in comment in Github. (Also possibly after each merge to develop). User flows: 10-20 specified user stories.

Didn’t do: Fastlane Jenkins setup pipeline (auto build Android/iOS).

Iteration 2
Goal Date: 2017-03-01
Description: Release 0.9.15 out and cut by Julien (4th person). Fastlane Jenkins setup pipeline (auto build Android/iOS) - Andy. Automated test: TBD. User flows: TBD.

**Comment by oskarth** [comment]
0.9.15 released a few days ago: https://blog.status.im/status-alpha-0-9-15-release-notes-63c41eb90eac

Shutting down this swarm per https://status-im.slack.com/archives/C8KB9DR60/p1521009678000295 (internal). Post-mortem TBD.

**Comment by oskarth** [comment]
Post-mortem notes: https://docs.google.com/document/d/1fX0gDdpKYBLopysfaKfegb7jQ6bZZnuTJVnbBnmovME/edit?usp=sharing
